src/pytorch_lightning/CHANGELOG.md

```diff
@@ -61,6 +61,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - When using multiple loggers, by default checkpoints and profiler output now get saved to the log dir of the first logger in the list ([14325](https://github.com/Lightning-AI/lightning/pull/14325))
 
 
+- Explicitly passing `train_dataloaders=None`, `val_dataloaders=None`, `dataloaders=None` or `datamodule=None` is officially no longer supported and will inform using a better error message than before ([14614](https://github.com/Lightning-AI/lightning/pull/14614))
```

Comment:
IMO this is a huge breaking change as described in https://github.com/Lightning-AI/lightning/issues/14602#issuecomment-1240607906.

Very often people write scripts with

```python
train_dataloader = None
if ...:
    train_dataloader = ...
trainer.fit(model, train_dataloader=train_dataloader)
```
(you had to update examples like this in our CI)

I don't think we can afford to break this usecase, sure, it can be error-prone, but it's a perfectly valid Python pattern when you want to pass arguments optionally.

I strongly think it should be a `PossibleUserWarning` for this reason.

Reply:
If we make it a warning, then another error will raise later on and completely overshadow any warning printed earlier. It won't help the user. 

The only other option I see is to change the message in "No trainining_step defined ..." to hint at the possibility that one of the dataloaders is empty. But this is very strange. Ideally we would want to tell the user directly what the issue is, not a list of possibilities.

Reply:
Another possibility would be to check at the same time whether there' a `*_dataloader` fallback. If there is, proceed normally. Otherwise, raise an error.

If we do this, we should also re-write the "No `train_dataloader()` method defined" error as it would be covered by this new error

Reply:
I have an alternative implementation here: #14637 

My brain is melting.

Reply:
Thanks! Much better